near-contract-standards/src/multi_token/core/core_impl.rs

```diff
@@ -0,0 +1,803 @@
+use near_sdk::borsh::{self, BorshDeserialize, BorshSerialize};
+use near_sdk::collections::{LookupMap, TreeMap, UnorderedMap, UnorderedSet};
+use near_sdk::json_types::U128;
+use near_sdk::{
+    assert_one_yocto, env, log, require, AccountId, Balance, BorshStorageKey, CryptoHash, Gas,
+    IntoStorageKey, PromiseOrValue, PromiseResult, StorageUsage,
+};
+
+use crate::multi_token::core::receiver::ext_mt_receiver;
+use crate::multi_token::core::resolver::{ext_mt_resolver, MultiTokenResolver};
+use crate::multi_token::core::MultiTokenCore;
+use crate::multi_token::events::{MtMint, MtTransfer};
+use crate::multi_token::metadata::TokenMetadata;
+use crate::multi_token::token::{Approval, ApprovalContainer, ClearedApproval, Token, TokenId};
+use crate::multi_token::utils::{
+    expect_approval, expect_approval_for_token, refund_deposit_to_account, Entity,
+};
+
+pub const GAS_FOR_RESOLVE_TRANSFER: Gas = Gas(15_000_000_000_000);
+pub const GAS_FOR_MT_TRANSFER_CALL: Gas = Gas(50_000_000_000_000 + GAS_FOR_RESOLVE_TRANSFER.0);
+
+const ERR_MORE_GAS_REQUIRED: &str = "More gas is required";
+const ERR_TOTAL_SUPPLY_OVERFLOW: &str = "Total supply overflow";
+const ERR_PREPAID_GAS_OVERFLOW: &str = "Prepaid gas overflow";
+const ERR_TOTAL_SUPPLY_NOT_FOUND_BY_TOKEN_ID: &str = "Total supply not found by token id";
+
+/// Implementation of the multi-token standard
+/// Allows to include NEP-245 compatible tokens to any contract.
+/// There are next traits that any contract may implement:
+///     - MultiTokenCore -- interface with transfer methods. MultiToken provides methods for it.
+///     - MultiTokenApproval -- interface with approve methods. MultiToken provides methods for it.
+///     - MultiTokenEnumeration -- interface for getting lists of tokens. MultiToken provides methods for it.
+///     - MultiTokenMetadata -- return metadata for the token in NEP-245, up to contract to implement.
+#[derive(BorshDeserialize, BorshSerialize)]
+pub struct MultiToken {
+    /// Owner of contract
+    pub owner_id: AccountId,
+
+    /// AccountID -> Near balance for storage.
+    pub accounts_storage: LookupMap<AccountId, Balance>,
+
+    /// The storage size in bytes for one account.
+    pub account_storage_usage: StorageUsage,
+
+    /// The storage size in bytes for one token.
+    pub storage_usage_per_token: StorageUsage,
+
+    /// How much storage takes every token
+    pub extra_storage_in_bytes_per_emission: StorageUsage,
+
+    /// Owner of each token
+    pub owner_by_id: TreeMap<TokenId, AccountId>,
```

Comment:
What is the reason of using TreeMap here?

Reply:
@Kirill-K-1 What do you propose is best data structure? Would you be fine with `UnorderedMap` so it can be iterated through?